File: android/app/src/main/java/com/choicecrafter/students/models/TaskStats.java
package com.choicecrafter.students.models;

public class TaskStats {
    private String attemptDateTime;
    private String timeSpent;
    private Integer retries;
    private Boolean success;
    private Boolean hintsUsed;
    private Double completionRatio;
    private Double scoreRatio;

    // No-argument constructor
    public TaskStats() {
    }

    public TaskStats(String attemptDateTime,
                     String timeSpent,
                     Integer retries,
                     Boolean success,
                     Boolean hintsUsed) {
        this(attemptDateTime, timeSpent, retries, success, hintsUsed, null, null);
    }

    public TaskStats(String attemptDateTime,
                     String timeSpent,
                     Integer retries,
                     Boolean success,
                     Boolean hintsUsed,
                     Double completionRatio) {
        this(attemptDateTime, timeSpent, retries, success, hintsUsed, completionRatio, null);
    }

    public TaskStats(String attemptDateTime,
                     String timeSpent,
                     Integer retries,
                     Boolean success,
                     Boolean hintsUsed,
                     Double completionRatio,
                     Double scoreRatio) {
        this.attemptDateTime = attemptDateTime;
        this.timeSpent = timeSpent;
        this.retries = retries;
        this.success = success;
        this.hintsUsed = hintsUsed;
        this.completionRatio = completionRatio;
        this.scoreRatio = scoreRatio;
    }

    public String getAttemptDateTime() {
        return attemptDateTime;
    }

    public void setAttemptDateTime(String attemptDateTime) {
        this.attemptDateTime = attemptDateTime;
    }

    public String getTimeSpent() {
        return timeSpent;
    }

    public void setTimeSpent(String timeSpent) {
        this.timeSpent = timeSpent;
    }

    public Integer getRetries() {
        return retries;
    }

    public void setRetries(Integer retries) {
        this.retries = retries;
    }

    public Boolean getSuccess() {
        return success;
    }

    public void setSuccess(Boolean success) {
        this.success = success;
    }

    public Boolean getHintsUsed() {
        return hintsUsed;
    }

    public void setHintsUsed(Boolean hintsUsed) {
        this.hintsUsed = hintsUsed;
    }

    public Double getCompletionRatio() {
        return completionRatio;
    }

    public void setCompletionRatio(Double completionRatio) {
        this.completionRatio = completionRatio;
    }

    public double resolveCompletionRatio() {
        if (completionRatio != null) {
            return Math.max(0.0, Math.min(1.0, completionRatio));
        }
        if (Boolean.TRUE.equals(success)) {
            return 1.0;
        }
        return 0.0;
    }

    public Double getScoreRatio() {
        return scoreRatio;
    }

    public void setScoreRatio(Double scoreRatio) {
        this.scoreRatio = scoreRatio;
    }

    public double resolveScoreRatio() {
        if (scoreRatio != null) {
            return Math.max(0.0, Math.min(1.0, scoreRatio));
        }
        return resolveCompletionRatio();
    }

    public boolean isCompleted() {
        if (Boolean.TRUE.equals(success)) {
            return true;
        }

        double completion = resolveCompletionRatio();
        if (completion >= 1.0) {
            return true;
        }

        if (completion > 0.0) {
            if (scoreRatio == null) {
                return true;
            }
            if (scoreRatio > 0.0) {
                return true;
            }
        }

        if (scoreRatio != null) {
            return scoreRatio > 0.0;
        }

        return false;
    }

    public boolean isFullyCorrect() {
        return resolveScoreRatio() >= 1.0;
    }
}

File: lib/models/task_stats.dart
class TaskStats {
  const TaskStats({
    this.attemptDateTime,
    this.timeSpent,
    this.retries,
    this.success,
    this.hintsUsed,
    this.completionRatio,
    this.scoreRatio,
  });

  final String? attemptDateTime;
  final String? timeSpent;
  final int? retries;
  final bool? success;
  final bool? hintsUsed;
  final double? completionRatio;
  final double? scoreRatio;

  factory TaskStats.fromMap(Map<String, dynamic> map) {
    return TaskStats(
      attemptDateTime: map['attemptDateTime'] as String?,
      timeSpent: map['timeSpent']?.toString(),
      retries: map['retries'] is num ? (map['retries'] as num).toInt() : null,
      success: map['success'] as bool?,
      hintsUsed: map['hintsUsed'] as bool?,
      completionRatio: _toDouble(map['completionRatio']),
      scoreRatio: _toDouble(map['scoreRatio']),
    );
  }

  Map<String, dynamic> toMap() {
    final map = <String, dynamic>{};
    if (attemptDateTime != null) {
      map['attemptDateTime'] = attemptDateTime;
    }
    if (timeSpent != null) {
      map['timeSpent'] = timeSpent;
    }
    if (retries != null) {
      map['retries'] = retries;
    }
    if (success != null) {
      map['success'] = success;
    }
    if (hintsUsed != null) {
      map['hintsUsed'] = hintsUsed;
    }
    if (completionRatio != null) {
      map['completionRatio'] = completionRatio;
    }
    if (scoreRatio != null) {
      map['scoreRatio'] = scoreRatio;
    }
    return map;
  }

  DateTime? attemptDateTimeParsed() {
    if (attemptDateTime == null || attemptDateTime!.isEmpty) {
      return null;
    }
    return DateTime.tryParse(attemptDateTime!);
  }

  double resolveCompletionRatio() {
    if (completionRatio != null) {
      return completionRatio!.clamp(0.0, 1.0);
    }
    if (success == true) {
      return 1.0;
    }
    return 0.0;
  }

  double resolveScoreRatio() {
    if (scoreRatio != null) {
      return scoreRatio!.clamp(0.0, 1.0);
    }
    return resolveCompletionRatio();
  }

  bool isCompleted() {
    if (success == true) {
      return true;
    }
    final completion = resolveCompletionRatio();
    if (completion >= 1.0) {
      return true;
    }
    if (completion > 0.0) {
      if (scoreRatio == null) {
        return true;
      }
      if (scoreRatio! > 0.0) {
        return true;
      }
    }
    if (scoreRatio != null) {
      return scoreRatio! > 0.0;
    }
    return false;
  }

  bool isFullyCorrect() {
    return resolveScoreRatio() >= 1.0;
  }
}

double? _toDouble(Object? value) {
  if (value is num) {
    return value.toDouble();
  }
  if (value is String) {
    return double.tryParse(value);
  }
  return null;
}
